Introduce Zmq and interrupt the main loop wait if zmq is setup and a block hash notification is received

RCasatta opened the pull request:
In a follow up MR we can also listen to transactions from zmq so that the mempool update already finds transactions and doesn't need to fetch most of them from the node (probably needs a tx cache #110 )

RCasatta committed:
introduce ZMQ listener thread and arg

shesek commented:
> In a follow up MR we can also listen to transactions from zmq ...

There's an WIP implementation of this I wrote awhile ago [in this branch](https://github.com/shesek/electrs/tree/202211-zmq).

It has basic working functionality for updating the mempool view with individual transactions as they arrive from zmq rather than doing full periodic updates, but is still incomplete and missing some important optimizations (in particular for updating the mempool scripthash history indexes, which was already inefficient but is in a much hotter code patch with this change).

> mempool update already finds transactions and doesn't need to fetch most of them from the node (probably needs a tx cache https://github.com/Blockstream/electrs/pull/110 )

This can be independent of a tx cache if you patch the mempool view directly with new transactions, instead of keeping them around but only applying them on the next periodic update.

---

My understanding is that in the current blockstream.info setup, there's a ZMQ client running alongside electrs that receives `blockhash` notifications from bitcoind and issues `SIGUSR1` signals to electrs. So in a way, we already are utilizing ZMQ to subscribe to block updates.

Given this, is there much of an advantage to moving the `blockhash` zmq listener into the electrs daemon?

I do see a significant advantage to using zmq for updating the mempool with individual transactions, both for performance and to keep the electrs mempool view more closely in-sync with bitcoind's, without the latency introduced by periodic updates.

However if that's our goal, perhaps we should aim straight for an MR that includes it too? I'm not sure that integrating zmq is worth it just for block updates, and it might be beneficial to design the zmq integration with mempool updates in mind from the get go.

RCasatta commented:
> Given this, is there much of an advantage to moving the blockhash zmq listener into the electrs daemon?

Yes, so we can get rid of a service in the infra written in go that listen to zmq and send signals, it simplify the infra to have this logic in electrs instead.

> However if that's our goal, perhaps we should aim straight for an MR that includes it too? 

We would need zmq thread similar to this for it? This seems a step in that direction and also useful to get rid of a part of the infra (the service with the purpose of sending signals). But I am fine also if we want to go with the branch instead (adding also the wakeup in the loop) if it can be done in a reasonable timeframe.

shesek commented:
> We would need zmq thread similar to this for it? .. But I am fine also if we want to go with the branch instead

Right, we do. If we do just blocks subscription, then there are some things I slightly prefer in the branch (using an optional feature flag and the ZmqSyncer encapsulation) but none are critical and we could go with this PR too.

> (adding also the wakeup in the loop) 

In the branch there no longer is a sync loop when ZMQ is enabled, it relies on the subscriptions instead of doing periodic updates:

https://github.com/shesek/electrs/blob/b071ea9c5d37a8855e1934137cbe19d07a65a890/src/bin/electrs.rs#L108-L118

> if it can be done in a reasonable timeframe.

Well define reasonable :) ZMQ-based mempool syncing is quite more involved than just block subscription so it would take longer to get right and require more throughout testing.

The upside is big and worth it imo, but if doing just block subscription to get rid of the zmq service is useful on its own then I wouldn't hold it up for that.

shesek commented on src/bin/electrs.rs:
```diff
@@ -123,14 +125,33 @@ fn run_server(config: Arc<Config>) -> Result<()> {
+        // We are going to wait 5 secs (50*100ms) if nothings happens.
+        // We are stopping if we receive a TERM or INT signal.
+        // We are interrupting the wait if we receive a block hash notification (if zmq enabled) or receiving a USR1 signal.
+        for _ in 0..50 {
```
Perhaps we could use crossbeam_channel's [`select!`](https://docs.rs/crossbeam-channel/latest/crossbeam_channel/macro.select.html) to wait for the first receiver that emits, instead of polling `block_hash_receive` on an interval?

Also, this could be made part of `Waiter::wait()`, with `accept_sigusr` changed into `accept_new_block_notif` to cover both cases.

RCasatta replied:
Changed to use `select!` in latest force push which is also rebased.
Note that since we want to use `select!` I removed previous latest commit, introduced `select!` so that one branch is the timeout, then added the zmq branch

RCasatta committed:
use select! instead of recv_deadline

this commit aim to achieve the same behaviour as before but using the
select! construct instead of recv_timeout because in the following
commits we want to add a case to the select!

RCasatta committed:
main loop listen also to block notify from zmq

shesek commented on src/new_index/zmq.rs:
```diff
@@ -0,0 +1,42 @@
+
+use crate::util::spawn_thread;
+
+pub fn start(url: &str, block_hash_notify: Option<Sender<BlockHash>>) {
```
Does `block_hash_notify` need to be an `Option`?

RCasatta replied:
974d78971fbfc9555e7a85fa32b2f91af6db8023

shesek commented on src/signal.rs:
```diff
@@ -26,34 +27,54 @@ fn notify(signals: &[i32]) -> channel::Receiver<i32> {
-        }
-    }
+    pub fn start() -> (Sender<BlockHash>, Waiter) {
+        let (block_hash_notify, block_hash_receive) = channel::bounded(1);
```
It feels more natural to me to create the `Sender` outside and accept it as an argument, but I don't feel strongly about this

RCasatta replied:
4edd741592ec00c3f1e33969ed26fc9f395f90e0

RCasatta committed:
zmq always accept a notifier

RCasatta committed:
Waiter accept receiver instead of returning sender

RCasatta committed:
rust format

shesek reviewed (approved):
Tested ACK a63c0634e741f1a696147f7b4ab5d80658ae2c2c

It could be nice if the `zmq` dependency was optional, but its good to go as-is too
